payjoin/src/receive/v2/mod.rs

```diff
@@ -472,6 +475,9 @@ pub struct PayjoinProposal {
 }
 
 impl PayjoinProposal {
+    // TODO hack to get multi party working. A better solution would be to allow extract_v2_req to be seperate from the rest of the v2 context
+    pub fn new(v1: v1::PayjoinProposal, context: SessionContext) -> Self { Self { v1, context } }
```

Comment:
I don't want to affect the v2 API with a hack. This could be merged as `pub(crate)` if possible or otherwise integrated into the state machine as your TODO suggests.